codalab/lib/codalab_manager_state.py

```diff
@@ -0,0 +1,201 @@
+
+    def set_last_check_version_datetime(self, timestamp):
+        raise NotImplementedError
+
+
+class CodaLabManagerJsonState(CodaLabManagerState):
+    def initialize_state(self):
+        if self.temporary:
+            self.state = {'auth': {}, 'sessions': {}}
+            return
+        # Read state file, creating if it doesn't exist.
+        if not os.path.exists(self.state_path):
+            write_pretty_json(
+                {
+                    'auth': {},  # address -> {username, auth_token}
+                    'sessions': {},  # session_name -> {address, worksheet_uuid}
+                },
+                self.state_path,
+            )
+        self.state = read_json_or_die(self.state_path)
+
+    @property
+    def state_path(self):
+        return os.getenv('CODALAB_STATE', os.path.join(get_codalab_home(), 'state.json'))
+
+    def get_auth(self, server, default={}):
+        return deepcopy(self.state["auth"].get(server, default))
+
+    def set_auth(
+        self, server, access_token, expires_at, refresh_token, scope, token_type, username
+    ):
+        self.state["auth"][server] = {
+            "token_info": {
+                "access_token": access_token,
+                "expires_at": expires_at,
+                "refresh_token": refresh_token,
+                "scope": scope,
+                "token_type": token_type,
+            },
+            "username": username,
+        }
+        self._save_json_state()
+
+    def delete_auth(self, server):
+        self.state["auth"].pop(server)
+        self._save_json_state()
+
+    def get_session(self, name, default={}):
+        return deepcopy(self.state["sessions"].get(name, default))
+
+    def set_session(self, name, address, worksheet_uuid):
+        self.state["sessions"][name] = {"address": address, "worksheet_uuid": worksheet_uuid}
+        self._save_json_state()
+
+    def get_last_check_version_datetime(self, default=None):
+        return deepcopy(self.state.get("last_check_version_datetime", default))
+
+    def set_last_check_version_datetime(self, timestamp):
+        self.state["last_check_version_datetime"] = timestamp
+        self._save_json_state()
+
+    def _save_json_state(self):
+        if self.temporary:
+            return
+        write_pretty_json(self.state, self.state_path)
+
+
+class CodaLabManagerSqlite3State(CodaLabManagerState):
+    def initialize_state(self):
+        if self.temporary:
+            self.connection = sqlite3.connect(":memory:")
+        else:
+            # Read state database, creating if it doesn't exist.
+            self.connection = sqlite3.connect(self.state_path)
+        self.connection.row_factory = sqlite3.Row
+        # Create the necessary tables, if they don't exist
+        with self.connection:
+            c = self.connection.cursor()
+            c.execute(
+                'CREATE TABLE IF NOT EXISTS auth (server TEXT UNIQUE PRIMARY KEY, access_token TEXT, expires_at REAL, '
```

Comment:
I'm worried sqlite is adding a lot of complexity (although I like the abstraction of `CodaLabManagerState`). How easy would it be to add a new column to state? (Would it be better to use an ORM or create migrations instead -- but that's even more complex, haha)

Are you sure we wouldn't be fine if we just added mutexes to make the process of reading from + writing to the JSON state file thread-safe? In that case, we could even create a new class, `CodaLabManagerJsonLockState`, that makes sure that the process of reading state & writing state is thread-safe.

Something like this:

```
acquire lock
read JSON state from disk
modify state and write to disk
release lock
```

Reply:
> How easy would it be to add a new column to state? (Would it be better to use an ORM or create migrations instead -- but that's even more complex, haha). 

Yeah, I thought about using an ORM / creating migrations, but I wanted to stay with something as simple as possible (this actually also motivated my use of sqlite, I'll mention more below). I don't think adding an extra column to the state is too difficult. The JSON state manager gets around this by just clobbering the existing state with whatever format it wants to write. We could do the same in the sqlite case, if we see that the existing table schema doesn't match what we expect. It's not ideal, but I think it gets around the issue fine without adding too much complexity.

Also, FWIW, blame seems to say that the state schema hasn't changed in several years. I don't want to say that it won't change in the future, but I historically it seems pretty infrequent—might be worth taking into consideration

> Are you sure we wouldn't be fine if we just added mutexes to make the process of reading from + writing to the JSON state file thread-safe?

No, I'm not sure :) I think that's a fine solution to the issue too. When I was first thinking of how to fix this issue, I was going to do something like this, but then I realized that sqlite has (1) basically already solved this problem, and (2) put a lot more effort into solving it robustly, so I opted for that instead. I thought that it'd be simpler to just import and use sqlite instead of building a lock system.